radix-engine/src/state_manager/staging.rs

```diff
@@ -1,189 +1,581 @@
+                for node_key in self.children_keys.iter() {
+                    self.dead_weight += Self::delete_recursive(
+                        &mut self.nodes,
+                        &mut self.total_weight,
+                        &new_root_key,
+                        visitor,
+                        node_key,
+                        0,
+                    );
+                }
 
-impl<'t, 's, S: ReadableSubstateStore> StagedSubstateStore<'t, 's, S> {
-    fn get_substate_recurse(&self, substate_id: &SubstateId, id: u64) -> Option<OutputValue> {
-        if id == 0 {
-            return self.stores.root.get_substate(substate_id);
-        }
+                let new_root = self.nodes.get(new_root_key).unwrap();
+                // Reparent to new_root node and delete it
+                self.children_keys = new_root.children_keys.clone();
+                for key in self.children_keys.iter() {
+                    let node = self.nodes.get_mut(*key).unwrap();
+                    node.parent_key = StagedSubstateStoreKey::RootStoreKey;
+                }
 
-        let node = self.stores.nodes.get(&id).unwrap();
-        if let Some(output) = node.outputs.get(substate_id) {
-            return Some(output.clone());
-        }
+                Self::remove_node(
+                    &mut self.nodes,
+                    &mut self.total_weight,
+                    visitor,
+                    &new_root_key,
+                );
 
-        self.get_substate_recurse(substate_id, node.parent_id)
+                // If the number of state changes that overlap with the self.root (dead_weight) store is greater
+                // than the number of state changes applied on top of it (total_weight), we recalculate the
+                // ImmutableStores in order to free up memory.
+                if self.dead_weight > self.total_weight {
+                    self.recompute_data();
+                }
+            }
+        }
     }
 }
 
-impl<'t, 's, S: ReadableSubstateStore> ReadableSubstateStore for StagedSubstateStore<'t, 's, S> {
-    fn get_substate(&self, substate_id: &SubstateId) -> Option<OutputValue> {
-        self.get_substate_recurse(substate_id, self.id)
-    }
+pub struct StagedSubstateStore<'t, S: ReadableSubstateStore> {
+    manager: &'t StagedSubstateStoreManager<S>,
+    key: StagedSubstateStoreKey,
 }
 
-impl<'t, 's, S: ReadableSubstateStore> WriteableSubstateStore for StagedSubstateStore<'t, 's, S> {
-    fn put_substate(&mut self, substate_id: SubstateId, output: OutputValue) {
-        if self.id == 0 {
-            panic!("Cannot write to root store");
-        } else {
-            let node = self.stores.nodes.get_mut(&self.id).unwrap();
-            node.outputs.insert(substate_id, output);
+impl<'t, S: ReadableSubstateStore> ReadableSubstateStore for StagedSubstateStore<'t, S> {
+    fn get_substate(&self, substate_id: &SubstateId) -> Option<OutputValue> {
```

Comment:
Does this fail if a substate is deleted in the child but not in the root?

Reply:
What do you mean by failing? Right now there is no implementation for deleting and the StagedStore reflects that. When we will have delete supported by the root store, this is easily fixed here by adding a Sentinel/Tombstone value.

Reply:
Ah cool kk. Can you add a comment regarding to implement Tombstoning here when required?